File: development/NET/samples/server/Session/VB/MyDaAddressSpaceElement.vb
Imports System
Imports System.Collections
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace Session
	Public Class MyDaAddressSpaceElement
		Inherits DaAddressSpaceElement

#Region "Constructors"
		'------------------------------

		Public Sub New(ByVal anItemID As String, ByVal aName As String, ByVal anUserData As UInt32, ByVal anObjectHandle As UInt32, ByVal aParentHandle As UInt32)
			MyBase.New(anItemID, aName, anUserData, anObjectHandle, aParentHandle)
		End Sub

		Public Sub New()
		End Sub

		'--
#End Region

#Region "Private Attributes"
		'-------------------------------

		Private m_properties As New Hashtable

		'--
#End Region

#Region "Public Methods"
		'---------------------

		''' <summary>
		''' Get elements property value data
		''' </summary>
		Public Sub GetPropertyValue(ByVal aRequest As DaRequest)
			Dim qualityGood As EnumQuality
			qualityGood = [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD)

			If aRequest.PropertyId = EnumPropertyId.ITEM_DESCRIPTION Then
				aRequest.Value = New ValueQT("description", qualityGood, DateTime.Now)

				aRequest.Result = EnumResultCode.S_OK
			Else
				aRequest.Result = EnumResultCode.E_NOTFOUND
			End If
		End Sub

		Public Overloads Overrides Function QueryProperties(ByRef aPropertyList As ArrayList) As Integer
			If m_properties.Count > 0 Then
				aPropertyList = New ArrayList
				aPropertyList.AddRange(m_properties.Values)
			Else
				aPropertyList = Nothing
			End If
			Return EnumResultCode.S_OK
		End Function


		Public Function AddProperty(ByVal aProperty As DaProperty) As Integer
			If Not aProperty Is Nothing Then
				m_properties.Add(aProperty.Id, aProperty)
				Return EnumResultCode.S_OK
			Else
				Return EnumResultCode.S_FALSE
			End If
		End Function

		Public Overloads Overrides Function QueryCacheValue(ByVal aSession As DaSession, ByRef aCacheValue As ValueQT) As Integer

			If IoMode = EnumIoMode.POLL_OWNCACHE Then

				Dim session As MySession = CType(aSession, MySession)
				If Not session Is Nothing Then
					Return session.GetCacheValue(aCacheValue)
				Else
					Return EnumResultCode.E_FAIL
				End If
			Else
				Return MyBase.GetCacheValue(aCacheValue)
			End If
		End Function

		Public Overloads Overrides Sub Change(ByVal isActive As Boolean, ByVal anUpdateRate As Integer)
			MyBase.Change(isActive, anUpdateRate)
			Console.m_clientChanged.[Set]()
		End Sub

		'--
#End Region

	End Class
End Namespace

File: development/NET/samples/server/Session/CS/MyDaAddressSpaceElement.cs
using System;
using System.Collections;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace Session
{
	public class MyDaAddressSpaceElement : DaAddressSpaceElement
	{

		#region Constructors
		//------------------------------

		public MyDaAddressSpaceElement(
			string anItemID,
			string aName,
			uint anUserData,
			uint anObjectHandle,
			uint aParentHandle)
			:
			base(
			anItemID,
			aName,
			anUserData,
			anObjectHandle,
			aParentHandle)
		{
		}   //	end constructor

		public MyDaAddressSpaceElement()
		{
		}   //	end constructor

		//--
		#endregion

		#region Private Attributes
		//-------------------------------

		private Hashtable m_properties = new Hashtable();

		//--
		#endregion

		#region Public Methods
		//---------------------

		/// <summary>
		/// Get elements property value data
		/// </summary>
		public void GetPropertyValue(DaRequest aRequest)
		{
			if (aRequest.PropertyId == (int)EnumPropertyId.ITEM_DESCRIPTION)
			{
				aRequest.Value = new ValueQT(
					"description",
					EnumQuality.GOOD,
					DateTime.Now);

				aRequest.Result = EnumResultCode.S_OK;
			}
			else
			{
				aRequest.Result = EnumResultCode.E_NOTFOUND;
			}   //	end if ... else
		}   //	end GetPropertyValue

		public override int QueryProperties(out ArrayList aPropertyList)
		{
			if (m_properties.Count > 0)
			{
				aPropertyList = new ArrayList();
				aPropertyList.AddRange(m_properties.Values);
			}
			else
			{
				aPropertyList = null;
			}   //	end if ... else
			return (int)(EnumResultCode.S_OK);
		}   //	end QueryProperties


		public int AddProperty(DaProperty aProperty)
		{
			if (aProperty != null)
			{
				m_properties.Add(
					aProperty.Id,
					aProperty);
				return (int)EnumResultCode.S_OK;
			}   //	end if
			else
			{
				return (int)EnumResultCode.S_FALSE;
			}   //	end if...else
		}   //	end AddProperty

		public override int QueryCacheValue(DaSession aSession, ref ValueQT aCacheValue)
		{

			if (IoMode == EnumIoMode.POLL_OWNCACHE)
			{

				MySession session = aSession as MySession;
				if (session != null)
				{
					return session.GetCacheValue(ref aCacheValue);
				}
				else
				{
					return (int)EnumResultCode.E_FAIL;
				}   //	end if ... else				
			}
			else
			{
				return base.GetCacheValue(ref aCacheValue);
			}   //	end if
		}   //	end QueryCacheValue

		public override void Change(bool isActive, int anUpdateRate)
		{
			base.Change(isActive, anUpdateRate);
			Console.m_clientChanged.Set();
		}   //	end Change

		//--
		#endregion

	}   //	end class MyAddressSpaceElement
}   //	end namespace
